### PR TITLE
tests: add repository level integration tests for kfp-api, kfp-viz, kfp-schedwf, and kfp-viewer

### PR DESCRIPTION
**Pull request by DnPlas**
In this PR:

* ci: run tox command with --bundle option

The `--bundle` option is now pointing at `tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2`, which contains a bundle definition for deploying kfp-operators as well as its dependencies.

*   tests: add integration tests at repo level for generic install
    
This commit adds repository level integration tests
    for improving the coverage of all kfp-components, except kfp-profile-controller.
    The components that are now covered are: kfp-api, kfp-viz, kfp-schedwf,
    and kfp-viewer.
This commit also removes the existing `test_bundle.py` and `test_pipelines.py` in favour of the new test suite.

*    tests: simplify and extract bundle management logic
    
This commit simplifies the bundle rendering and applying logic, and
    abstracts it from the test files so it can be reused.
    This reusable set of helper methods leverage pytest-operator's
    render_bundle()[1] and COS' cli_deploy_bundle()[2].

#### TODO:
- [x] Remove the skip from the test cases that are failing because of Authentication
- [ ] Fix the `wait_run_for_completion` error in `test_create_and_monitor_run`. Not fixing, see https://github.com/canonical/kfp-operators/issues/244
- [ ] Fix the disabling not working in `test_create_and_monitor_recurring_run`. Not fixing, see #245 
- [x] Document how to test (locally, GH CI, against edge, against a version) and how to update tests
- [x] Remove RBAC workaround
- [x] Remove authentication

Links:
    [1] https://github.com/charmed-kubernetes/pytest-operator/blob/main/pytest_operator/plugin.py#L1232
    [2] https://github.com/canonical/cos-lite-bundle/blob/main/tests/integration/helpers.py#L19

### CODE REVIEW

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (ca-scribner) [inline tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:6]: Are the revisions here intentional?  The file name says "1.7_stable", but with the revisions here this wont actually keep up with 1.7's stable but instead be whatever 1.7 stable was when the file was created.  

I wonder if this is an export from the `juju export-bundle` command, and some attributes need to be removed before including it here?

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (ca-scribner) [inline tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:21]: (Ignore this if there's more to come that'll benefit from these dependencies)

Looking at the tests now, I don't see enough value to warrant including dex-auth, istio, and oidc in these tests.  I see that we do get to test logging in with an authenticated session (`auth_session`, etc), but that feels like something we still need to test at the kubeflow-bundle level and something that generally wont break because of a change in kfp.  Outside that, I'm missing what else it adds to the test outcomes

File: tests/integration/test_kfp_generic_installation.py
Comment (ca-scribner) [inline tests/integration/test_kfp_generic_installation.py:313]: this might make for a nice helper like:
```
async def get_address_from_unit(model, unit_name: str):
  app_name = "".join(unit_name.split("/")[:-1])
  status = await model.get_status()
  units = status["applications"][app_name]["units"]
  return units[unit_name]["address"]

def test_something(ops_test):
  ...
  url = await get_address_from_unit(ops_test.model, "kfp-viz/0")
```

**Review by ca-scribner — COMMENTED** [review]
Overall things looks good!  Some high-level notes:
* I have some concerns about avoiding unnecessary charm building when locally running these tests (see comments on that in the review)
* I wonder if, without having kubeflow profiles deployed, the behaviour of kfp is different enough that it leaves gaps in testing.  I'm thinking about for testing pipeline execution, etc, real users run these from the perspective of a kubeflow user (which changes the permissions, etc, of the execution).  I don't know if anything is *wrong* here, but I am not confident that we aren't omitting something important either

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:6]: It is an export, that's correct, I did not bother removing the revisions and other attributes, but I can do it. Good catch.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:32]: We want this to configure Dex's static username and password, the ones used in the test are for authenticating (so equivalent of you introducing your passwd/username). I could read the values in this bundle definition and use them in the test file, if that sounds better.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:89]: Yeah, the case that you mentioned is not covered here. Another thing I've seen in other charms ([cos-lite](https://github.com/canonical/cos-lite-bundle/blob/main/bundle.yaml.j2)) is that they have a condition per charm, so they will render only iff the context contains a local charm path, otherwise, they will use a charm from CH.
I did not want to do that because it would've been a lot more configuration on the test side (helpers). If you think it is necessary, though, I can work on it a bit more.

File: tests/integration/data/kfp_latest_edge.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_latest_edge.yaml.j2:1]: No, I left it there in case someone wanted to test against latest/edge. My idea is to have a collection of bundle files that can be used for testing. I want to document that as well, let me add it to the TODO list.

File: tests/integration/pipelines/create_sample_pipeline.py
Comment (DnPlas) [reply tests/integration/pipelines/create_sample_pipeline.py:5]: (:

File: tests/integration/pipelines/create_sample_pipeline.py
Comment (DnPlas) [reply tests/integration/pipelines/create_sample_pipeline.py:1]: It is tooling, I wanted to keep it because sometimes we compile a pipeline and we use it in some test cases, but then it gets outdated and no-one seems to know where it came from. I will document how these are used, added to my TODO list above.

File: tests/integration/pipelines/execution_order.py
Comment (DnPlas) [reply tests/integration/pipelines/execution_order.py:1]: It's not used, see my [other comment](https://github.com/canonical/kfp-operators/pull/238#discussion_r1252374690)

File: tests/integration/pipelines/create_sample_pipeline.py
Comment (DnPlas) [reply tests/integration/pipelines/create_sample_pipeline.py:1]: Actually, this one is not used at all for creating the sample pipeline that is used in the test, so I will remove the file. I added it by mistake.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py:80]: Yeah, at first I added them as helpers, but I kept using them so often, that I thought they would fit better as fixtures. I get the feedback, though, and I have no arguments against moving away from fixtures. I will update.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py:198]: > Also, I realise you inherited this top level variable pattern from the previous test suite, and I'm scared to look but I bet the git blame points at me for it :)

I did inherit the top level variable from the previous test suite.

> Do we need to define the GENERIC_BUNDLE_CHARMS here? Or could we infer it from the charms dir? There's maybe an edge case where we add a charm in the repo that isn't part of these tests, but realistically I don't think that'll happen

I wanted to have a definition w/o the profiles-controller to keep this installation as "generic" as possible, that is, without multi-user isolation. In this case, inferring from the charms dir would've meant including the charm I don't want.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py:284]: The thing is that the viewer server's only job is to provide the CRD (VIewer) and make sure the CRs can be created, watched, listed, deleted, etc. The visualization component (below), could have something else for the Viewer. I will think of something.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py:313]: Agreed, will update.

**Comment by DnPlas** [comment]
> Overall things looks good! Some high-level notes:
>     * I wonder if, without having kubeflow profiles deployed, the behaviour of kfp is different enough that it leaves gaps in testing.  I'm thinking about for testing pipeline execution, etc, real users run these from the perspective of a kubeflow user (which changes the permissions, etc, of the execution).  I don't know if anything is _wrong_ here, but I am not confident that we aren't omitting something important either

Thanks for the review @ca-scribner I plan to submit another PR that adds a multi-user installation, where kfp-profile-controller is added and tested. The other test suite will include kfp-profile-controller existing tests plus the KFP API focused tests in this PR. Does that make more sense?

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:21]: Yes dex, oidc, and istio are there for authentication purposes only, this is to be able to connect the KFP Client to the KFP Server. If you have an idea on how to do this differently, let me know.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (kimwnasptd) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:32]: A note here that we could avoid Dex/AuthService in testing altogether by:
1. Port-forwarding the KFP K8s Service to localhost
2. Updating the client to use that host, but manually setting the userid header (if needed)

We had done this in upstream KF tests
https://github.com/kubeflow/manifests/blob/master/tests/e2e/hack/proxy_pipelines.sh#L4
https://github.com/kubeflow/manifests/blob/master/tests/e2e/mnist.py#L48-L51

But if you've tested and are more comfortable with the current implementation we can do the above in a follow-up improvement PR later on

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2:21]: The other option I considered was [this](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub), which uses `PodDefaults` for mounting the `ServiceAccount` volume on all pods this test suite would use, but to me it was a bit more useful in cases where we use the REST API.

**Comment by kimwnasptd** [comment]
@DnPlas @ca-scribner regarding the discussion point above on multi-user mode, the change in the multi-user isolation manifests (from upstream) are:
1. Istio VirtualServices, for allowing only Istio IngressGateway to set the user header https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/istio-authorization-config.yaml
2. Making the server have authorization checks https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/pipelines-ui/deployment-patch.yaml#L25-L30

So i'd suggest that it's OK not to test multi-user here and have the multi-user story for the bundle testing, which will be creating a Pipeline in the user's namespace etc

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: I have updated the bundle definition removing all the extra information that is not needed.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py]: I have used a slightly different approach here, let me know what you think.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: Updated.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: Updated.

File: tests/integration/pipelines/create_sample_pipeline.py
Comment (DnPlas) [reply tests/integration/pipelines/create_sample_pipeline.py:1]: Removed the file and left `sample_pipeline_execution_order.py` as it's the one used for compiling the sample pipeline.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (ca-scribner) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: I can understand this flexibility wasn't part of the initial scope, but since this refactor is moving some existing integration tests away from the charm and to the bundle I think we're harming the developer experience unless single-charm building is enabled.  Looking at the PRs to charms in this repo they're usually edits to a single charm not the entire bundle, so the typical development experience is focused on editing one charm in relation to others that are unchanged.  Since building the whole bundle locally takes ~1hr (I might be misremembering, but at least 30 min) the development experience after this PR merges sounds frustrating imo

File: tests/README.md
Comment (ca-scribner) [inline tests/README.md:28]:
```suggestion
* `integration/bundles`: contains a set of bundle definition Jinja2 templates that can be passed to the `tox` command to run integration tests against different configurations for the kfp-operators. The CI runs the latest stable version of this bundle definition by default.
```

File: tests/integration/test_kfp_generic_installation.py
Comment (ca-scribner) [reply tests/integration/test_kfp_generic_installation.py:255]: Are you sure about this?  kfp-viewer deploys the `viewer-crd-controller` image, which I think is [this source code](https://github.com/kubeflow/pipelines/blob/master/backend/src/crd/controller/viewer/main.go).  I see it creating services and deployments, plus some other stuff.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (ca-scribner) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: I'll continue the conversation in [this thread](https://github.com/canonical/kfp-operators/pull/238#discussion_r1252260460) as I think they're basically the same.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (ca-scribner) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: (bringing conversation over from [this thread](https://github.com/canonical/kfp-operators/pull/238#discussion_r1252298756) too, which is nearly the same)

Thinking about what functionality the dependencies bring:
* istio pilot+gateway: 
  * the external gateway url (so we don't need a port-forward)
  * the additional `kubeflow-userid` header **iff** we also have dex+auth
* dex+auth:
  * the requirement that users authenticate
  * the `kubeflow-userid` header

From that, I think we have a few options:
1. use dex+auth+istio
  * we access through the external url and log in via auth (like this pr already had)
  * `kubeflow-userid` is assigned automatically for us
2. omit dex/auth, but keep istio
  * we do not get `kubeflow-userid` automatically, so we need to add the header ourselves
  * we can connect to the external url (but this might remove the `kubeflow-userid` header)
3. omit dex/auth/istio
  * we can connect to kfp directly via port-forwarding
  * we must add the kubeflow-userid-header

I don't know which is better really.  (1) is closest to charmed kubeflow, feels like it adds more dependencies than the value we get from them.  (2) means we dont have to port-forward, but otherwise doesn't add much.  (3) makes connecting harder.  They're all bad in their own way :D

Comment (ca-scribner) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: But maybe I'm over complicating things, at least for this PR.  It feels heavy to deploy istio+dex+auth just so that we also need to log in, but then otherwise use kfp without caring about those tools.  But on the other side, it is simpler to implement because we're used to it.  So I trust your judgement in picking which feels best

File: tests/README.md
Comment (ca-scribner) [inline tests/README.md:41]: I think this is required for all builds (`ops_test.build_charm()` [calls `charmcraft pack`](https://github.com/charmed-kubernetes/pytest-operator/blob/9e0c8f3b4feb848e097391317c5a0a09d43bd3f4/pytest_operator/plugin.py#L965)

File: tests/integration/pipelines/create_sample_pipeline.py
Comment (ca-scribner) [reply tests/integration/pipelines/create_sample_pipeline.py:1]: perfect.  as long as there's instructions I'm +1

File: tests/README.md
Comment (ca-scribner) [reply tests/README.md:41]: ohhh yeah I thought it was talking about my computer vs something in GH.  Now I see what you mean though

File: tests/integration/test_kfp_generic_installation.py
Comment (ca-scribner) [reply tests/integration/test_kfp_generic_installation.py]: lgtm!

**Comment by ca-scribner** [comment]
@kimwnasptd are you sure that's all that changes in the upstream manifests?  That `installs/multi-user` has a lot more content.  For example, the [`params.env`](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/api-service/params.env) that enables `MULTIUSER=true` and `DEFAULTPIPELINERUNNERSERVICEACCOUNT=default-editor`.  I believe the service account part is because with `MULTIUSER=true` kfp runs things as a user not in the control plane.

**Comment by DnPlas** [comment]
After a discussion with @ca-scribner and @kimwnasptd, it was decided that the test suite provides good coverage for the functional part of kfp-operators without the strict need for kfp-profiles-controller. 
For communicating with the KFP API Server, it was agreed to use a test `Profile`, which is controlled by the `kubeflow-profiles` and assigned clusterRoles by the `kubeflow-roles` charms. These two will be included in the bundle that's deployed for testing.
Lastly the integration tests are failing due to:
1. `wait_for_idle` is not waiting for all units to be active and idle, which affects all tests cases as the units (specially the workloads) are not ready to work.
2. The `kubeflow-roles` seems to not be providing enough permissions to the test `Profile`, returning `HTTP response body: {"error":"Failed to authorize with API resource references: Failed to authorize with API resource references: PermissionDenied: User 'prof
ilename' is not authorized with reason`. The latter should be investigated.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: This conversation does not apply anymore to the current testing setup. Feel free to re-open if further discussion is needed.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py]: Please take a look at the readme, and let me know if there's anything missing there.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: This conversation does not apply anymore to the current testing environment. Feel free to re-open if needed.

File: tests/README.md
Comment (DnPlas) [reply tests/README.md]: I have re-phrased it a bit, let me know what you think.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: I think the developer experience hasn't really changed between the current state of the repo (building and deploying ALL charms) and the proposal in this PR, but let me know if it has. 
Now that of it, the kind of optimisation you mention (changes in a single charm would trigger builds only for that charm) belong to a different effort, which we can combine with cached builds, and really improve the CI times. For now, I think it is better keep the scope of this PR as is.

File: tests/integration/test_kfp_functional.py
Comment (DnPlas) [inline tests/integration/test_kfp_functional.py:139]: I have added an issue: https://github.com/canonical/kfp-operators/issues/244

File: tests/integration/test_kfp_functional.py
Comment (DnPlas) [inline tests/integration/test_kfp_functional.py:181]: I have added an issue #245

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (ca-scribner) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: From the spec, I thought the plan was removing kfp-api's charm integration tests and instead using the bundle tests.  I don't see that change in this PR, but I'm guessing from the spec that it is planned.  Is that the case?  Using these bundle tests as kfp-api's integration test is the regression I'm thinking of

File: tests/README.md
Comment (ca-scribner) [reply tests/README.md]: lgtm!

File: tests/integration/test_kfp_functional.py
Comment (ca-scribner) [reply tests/integration/test_kfp_functional.py:139]: What state is this run and its k8s objects in when it is hanging?  Is this something that would eventually resolve (eg: just needs more time/resources) or something that is broken?  I get the feeling you're saying it just needs more time/resources, but these same symptoms could be because it just isn't running the job properly

Comment (DnPlas) [reply tests/integration/test_kfp_functional.py:140]: It is running, it will have a state, but it doesn't seem to be able to finish within the timeout limits. If I increase the timeout, I then hit another problem with the `port-forward` losing connection to the svc it is forwarding.

What I will do is retrieve that state, check that it is running at least until we figure out a solution for the issue. Would that be okay?

File: tests/integration/conftest.py
Comment (ca-scribner) [inline tests/integration/conftest.py:131]: This cannot be turned off:
* `tox -e bundle-integration`: results in `build=True`
* `tox -e bundle-integration -- --build`: results in `build=True`
* `tox -e bundle-integration -- --build=False`: results in `build="False"` (Note, it is a **string** of False, not the boolean)
  * This will evaluate as True [here](https://github.com/canonical/kfp-operators/pull/238/files/f602d6a13bfb39e2199c153c78d98584b885a8d1#diff-2539489e7622d481095b67a9cf5cd2e3366be6816e2ad192c9996de42c1d7487R22)

I think there is a different argument type here that you could use that lets someone do `--build` or `--no-build` to control the boolean.

File: tests/integration/test_kfp_functional.py
Comment (ca-scribner) [reply tests/integration/test_kfp_functional.py:140]: My worry is that it is not finishing because of a configuration problem within kfp-api.  I think we can misconfigure kfp-api in a way where it can start jobs, but cannot keep up to date on their progress (so they'll perpetually look like they're running).  I don't think the runner is simply overloaded here, I think what you're seeing is your tests correctly showing you a real bug

Comment (ca-scribner) [reply tests/integration/test_kfp_functional.py:140]: I'm curious what the Pods, deployments, etc from the jobs say (`kubectl get` and `kubectl describe` on pod, deployment, and all the pipelines-related objects for a run)

Comment (kimwnasptd) [reply tests/integration/test_kfp_functional.py:140]: @DnPlas since we have a follow-up issue, could we expose some logs from Pods and add them on the other issue and unblock this PR?

While I'd also prefer that we have the answer on why it times out only in the GH Runner at the same time let's continue with merging this first PR.

Comment (ca-scribner) [reply tests/integration/test_kfp_functional.py:140]: @dnplas and I discussed.  Agreed that this should be addressed as a separate issue

Comment (DnPlas) [reply tests/integration/test_kfp_functional.py:140]: @kimwnasptd there is a GH issue for tracking this #244, I will be adding logs there. @ca-scribner and I have discussed this offline and have confirmed that this behaviour is only observed on the GH runners, when running on a machine with more resources, the `Run` actually gets `status == 'Succeeded'`.
I will resolve this conversation, as the issue should cover all details, but feel free to re-open if further discussion is needed.

**Review by ca-scribner — APPROVED** [review]
lgtm! really great work

File: tests/integration/conftest.py
Comment (DnPlas) [reply tests/integration/conftest.py]: Updated to use `--no-build`, thanks for the suggestion.

File: tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2
Comment (DnPlas) [reply tests/integration/data/kfp_1.7_stable_generic_install.yaml.j2]: No, those will remain there.

File: tests/integration/test_kfp_generic_installation.py
Comment (DnPlas) [reply tests/integration/test_kfp_generic_installation.py:1]: Thanks for the suggestion, I did it.

**Pull request merged**
